### PR TITLE
Upsert keys in .env for benchmarks auth/init instead of appending

### PR DESCRIPTION
**Pull request by kaggle-agent**
`kaggle b auth` and `kaggle b init` opened `.env` in append mode, so
every rerun stacked a fresh copy of `MODEL_PROXY_*` (and `LLM_*`)
onto the bottom of the file. Two problems: the user's `.env` grew
without bound, and any value the user had hand-edited got silently
overridden by the appended copy below it (python-dotenv reads the
last occurrence). Switch to `dotenv.set_key(..., quote_mode="never")`
so each managed key is refreshed in place, preserving comments, blank
lines, and unrelated keys.

Migration note: users with stacked duplicates from prior runs keep
those lines, but `set_key` rewrites every occurrence to the fresh
value, so no stale token lingers in the file after the first rerun
on this code.

---

Task: [limagoog-20260529183140-af82f3a6](https://agents.dev.kaggle.net/tasks/limagoog-20260529183140-af82f3a6)
Context: https://chat.kaggle.net/kaggle/pl/46qfted3gbyqdkcpt447cusich

### CODE REVIEW

File: src/kaggle/api/kaggle_api_extended.py
Comment (dolaameng) [inline src/kaggle/api/kaggle_api_extended.py:1]: Always run `hatch run lint:fmt` and `hatch run lint:all` as pre-submit.

Comment (kaggle-agent) [reply src/kaggle/api/kaggle_api_extended.py:1]: Ran `hatch run lint:fmt` and `hatch run lint:all`. Black reformatted two adjacent-string-literal blocks in `test_benchmarks_cli.py` onto single lines (changes are formatting-only, no semantic difference). `lint:all` then reports `Success: no issues found in 21 source files`. Will keep this in the pre-submit loop going forward.

File: src/kaggle/api/kaggle_api_extended.py [hunk 1]
```diff
@@ -7193,15 +7193,47 @@ def _write_benchmarks_env(self, env_vars, no_confirm, env_file, quiet=False):
...
         if not quiet:
             print(f"Environment variables have been written to {env_file_abs}.")
         return True
 
+    # Matches ``KEY=`` at the start of an env-file line, allowing leading
+    # whitespace and an optional ``export`` prefix.
+    _ENV_LINE_RE = re.compile(r"^[ \t]*(?:export[ \t]+)?([A-Za-z_][A-Za-z0-9_]*)[ \t]*=")
+
+    @classmethod
+    def _upsert_env_file(cls, env_file_abs, env_vars):
```
Comment (dolaameng) [inline src/kaggle/api/kaggle_api_extended.py]: Shall we use existing library like `python-dotenv` instead of doing it from scratch?

File: src/kaggle/test/test_benchmarks_cli.py [hunk 2]
```diff
@@ -2111,6 +2111,200 @@ def test_appends_to_existing_file(self, api, mock_token, capsys, tmp_path):
         assert "LLM_DEFAULT=google/gemini-3-flash-preview\n" in content
 
 
+# ============================================================
```
Comment (dolaameng) [inline src/kaggle/test/test_benchmarks_cli.py:2114]: These are too much new test code. We want good coverage but less redundancy. e.g.,  `test_writes_to_empty_file`,` test_rerunning_auth_replaces_existing_keys``test_preserves_user_added_keys_and_comments` are all redundant. Specially if we decide to adopt `python-dotenv`, which already provides tested expected behavior. Can simplify these?

Comment (kaggle-agent) [reply src/kaggle/test/test_benchmarks_cli.py:2114]: Trimmed the new tests significantly now that `python-dotenv` owns the line-by-line semantics. Deleted the whole `TestUpsertEnvFile` class (8 tests) and the redundant `test_rerunning_auth_replaces_existing_keys` integration test. Kept only 3 integration tests under `TestBenchmarksUpsert` that verify the user-facing contract through the actual CLI: stale token gets replaced on rerun, user-added keys + comments survive, and `init` reruns are idempotent across all six managed keys. Net diff is now ~55 lines of test code instead of ~210.

File: src/kaggle/api/kaggle_api_extended.py (hunk 1 above)
Comment (kaggle-agent) [reply src/kaggle/api/kaggle_api_extended.py]: Switched to `python-dotenv`. Added `python-dotenv` to dependencies, removed the bespoke `_upsert_env_file` helper + `_ENV_LINE_RE`, and replaced the write block with a small loop calling `dotenv.set_key(env_file_abs, key, value, quote_mode="never")`. `quote_mode="never"` preserves the existing unquoted output format (so the existing auth/init assertions like `MODEL_PROXY_URL=https://...` still hold).

File: src/kaggle/test/test_benchmarks_cli.py [hunk 3]
```diff
@@ -2111,6 +2111,65 @@ def test_appends_to_existing_file(self, api, mock_token, capsys, tmp_path):
...
+# Upsert behavior (auth/init reruns)
+# ============================================================
+
+
+class TestBenchmarksUpsert:
```
Comment (dolaameng) [inline src/kaggle/test/test_benchmarks_cli.py:2119]: Shall we add a test for the migration-from-duplicates scenario that's the main user-facing promise.

Comment (kaggle-agent) [reply src/kaggle/test/test_benchmarks_cli.py:2119]: Added `test_first_rerun_after_upgrade_refreshes_stale_duplicates`. It seeds a `.env` with three stacked `MODEL_PROXY_API_KEY=` lines (mimicking what the old append-based CLI would have left behind), runs `kaggle b auth` once, and asserts (a) no stale value survives, (b) `dotenv_values()` — which is what consumers use — resolves to the fresh token, and (c) unrelated keys are untouched. Worth noting: `dotenv.set_key` refreshes *every* duplicate occurrence, not just the first, so the migration story is stronger than I'd written in the PR description — the live value is correct after the first rerun even though the duplicate lines themselves remain.

**Review by dolaameng — APPROVED** [review]
LGTM

**Review by nl917 — APPROVED** [review]
LGTM! Thanks

File: src/kaggle/api/kaggle_api_extended.py [hunk 4]
```diff
@@ -7193,10 +7194,11 @@ def _write_benchmarks_env(self, env_vars, no_confirm, env_file, quiet=False):
             if not self.confirmation(f"write these settings to {os.path.basename(env_file_abs)}", default_to_yes=True):
                 return False
 
-        with open(env_file_abs, "a") as f:
-            f.write("\n")
-            for key, value in env_vars.items():
-                f.write(f"{key}={value}\n")
+        # Upsert via python-dotenv so reruns refresh in place rather than
```
Comment (dolaameng) [inline src/kaggle/api/kaggle_api_extended.py]: Is there comment still accurate? Try to keep it concise

Comment (kaggle-agent) [reply src/kaggle/api/kaggle_api_extended.py]: Trimmed the comment from 3 lines to 2. Kept only the non-obvious bit (why upsert vs append: `python-dotenv` picks the first occurrence, so duplicates would let an expired token shadow the fresh one) and dropped the redundant phrasing.

File: src/kaggle/api/kaggle_api_extended.py [hunk 5]
```diff
@@ -7193,10 +7194,10 @@ def _write_benchmarks_env(self, env_vars, no_confirm, env_file, quiet=False):
             if not self.confirmation(f"write these settings to {os.path.basename(env_file_abs)}", default_to_yes=True):
                 return False
 
-        with open(env_file_abs, "a") as f:
-            f.write("\n")
-            for key, value in env_vars.items():
-                f.write(f"{key}={value}\n")
+        # Upsert (not append): python-dotenv reads the first occurrence, so
```
Comment (dolaameng) [inline src/kaggle/api/kaggle_api_extended.py]: I don't think the comment is correct. The purpose of this PR is to keep the .env clean on user side, and also avoid conflicts with user manual editing. not because the new keys shadowing the old ones, because python-dotenv actually reads the last one for duplicates.

Fix this comment (or just say what it does briefly without these details), and also update your PR summary.

Comment (kaggle-agent) [reply src/kaggle/api/kaggle_api_extended.py]: You're right — `python-dotenv` reads the *last* occurrence, not the first (I verified: `dotenv_values` on `FOO=first\nFOO=middle\nFOO=last\n` returns `last`). So the actual motivation is file hygiene (no stacked duplicates on the user side) and not silently shadowing any value the user has hand-edited. Updated the inline comment to say exactly that, and corrected the test docstring on `test_first_rerun_after_upgrade_refreshes_stale_duplicates` (its assertions were already correct — `set_key` refreshes every occurrence, so no stale value lingers — but the prose had the same wrong framing). I'll update the PR summary in the next regeneration step.
